### PR TITLE
[POC] Add UserData to Heartbeat

### PR DESCRIPTION
**Pull request by vvcephei**
Prototype of adding an extra payload to the heartbeat protocol, with KIP-441 in mind.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java [hunk 1]
```diff
@@ -252,4 +252,5 @@
      */
     void wakeup();
 
+    void setHeartbeatCallbacks(ConsumerHeartbeatDataCallbacks callbacks);
```
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: Would actually either be added via config or mixed into the Assignor interface

File: clients/src/main/java/org/apache/kafka/common/requests/HeartbeatResponse.java [hunk 2]
```diff
@@ -57,6 +57,10 @@ public Errors error() {
         return Errors.forCode(data.errorCode());
     }
 
+    public ByteBuffer userDatas() {
```
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/common/requests/HeartbeatResponse.java]: This is the serialized map of MemberId -> UserData . I was reluctant to deserialize it eagerly for performance reasons. Not sure if this is reasonable or not.

I guess even if we want it to be lazy, we should encapsulate the deserialization and just cache the result.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 3]
```diff
@@ -222,7 +222,7 @@ public void testNormalHeartbeat() {
 
         // normal heartbeat
         time.sleep(sessionTimeoutMs);
-        RequestFuture<Void> future = coordinator.sendHeartbeatRequest(); // should send out the heartbeat
+        RequestFuture<Map<String,ByteBuffer>> future = coordinator.sendHeartbeatRequest(); // should send out the heartbeat
```
Comment (vvcephei) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Not sure if `coordinator.sendHeartbeatRequest()` is really a public API or not.

File: streams/src/test/java/org/apache/kafka/streams/integration/HeartbeatIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        @Override
+        public void leaderAllMemberUserDatas(Map<String, ByteBuffer> userDatas) {
+            for (Map.Entry<String, ByteBuffer> entry : userDatas.entrySet()) {
+                System.out.println(Instant.now() +" member["+entry.getKey()+"] data["+new String(entry.getValue().array(), StandardCharsets.UTF_8)+"]");
+            }
+        }
+    }
+
+    @Test
+    public void testHeartBeatCommunication() throws InterruptedException {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/HeartbeatIntegrationTest.java:245]: not really a test, just a little POC demonstration (doesn't terminate)

**Comment by vvcephei** [comment]
The basic pattern is that each member sends some payload with their heartbeats, then the group leader saves them internally and responds to the consumer leader's heartbeat request with a map of `MemberId->UserData` .

I guess we could also send the whole group's UserDatas in all the members' heartbeat responses, and this could be a way to generally broadcast information around the cluster, but it seems like an abuse of the protocol.

The way I'm thinking about it, the consumer leader would make use the of the members' payloads in allocation/rebalance decisions, so only the leader needs to get the information.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 5]
```diff
@@ -880,27 +884,38 @@ public void handle(LeaveGroupResponse leaveResponse, RequestFuture<Void> future)
...
+    synchronized RequestFuture<Map<String,ByteBuffer>> sendHeartbeatRequest() {
         log.debug("Sending Heartbeat request to coordinator {}", coordinator);
         HeartbeatRequest.Builder requestBuilder =
                 new HeartbeatRequest.Builder(new HeartbeatRequestData()
                         .setGroupId(rebalanceConfig.groupId)
                         .setMemberId(this.generation.memberId)
                         .setGroupInstanceId(this.rebalanceConfig.groupInstanceId.orElse(null))
-                        .setGenerationId(this.generation.generationId));
+                        .setGenerationId(this.generation.generationId)
+                        .setUserData(getHeartbeatData())
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Seems some changes in `HeartbeatRequest` was not included?

Comment (vvcephei) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: yeah, I didn't realize I'd edited a generated class.

**Pull request closed without merging**